### PR TITLE
[#10295] Allow using the informer interface, and trace down all types

### PR DESCRIPTION
**Pull request by ctron**
This change also evaluates implementations of `ResourceEventHandler`,
which is the corresponding interface for the informer. This allows to
use watchers or informers the same way.

Also it traverses through the type information, adding all required
types to the list of types enabling reflection. If only the direct class
is added, deserialization will fail, unless all the references classes
are manually annotated with @RegisterForReflection.

Signed-off-by: Jens Reimann <jreimann@redhat.com>

### CODE REVIEW

**Comment by geoand** [comment]
I'll be looking at this tomorrow

**Comment by geoand** [comment]
Thanks a lot for this @ctron.

Can you add a description of what exactly the type expansion is doing and why it's needed? Perhaps with an example?
I would really like to know what's going on because we would really like to avoid adding to many things to reflection configuration.

File: extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java [hunk 1]
```diff
@@ -23,12 +30,25 @@
...
 public class KubernetesClientProcessor {
 
     private static final DotName WATCHER = DotName.createSimple("io.fabric8.kubernetes.client.Watcher");
+    private static final DotName RESOURCE_EVENT_HANDLER = DotName
+            .createSimple("io.fabric8.kubernetes.client.informers.ResourceEventHandler");
     private static final DotName KUBERNETES_RESOURCE = DotName
             .createSimple("io.fabric8.kubernetes.api.model.KubernetesResource");
+    private static final DotName KUBERNETES_RESOURCE_LIST = DotName
+            .createSimple("io.fabric8.kubernetes.api.model.KubernetesResourceList");
+    private static final Set<String> EXCLUDED_CLASSES;
```
Comment (geoand) [inline extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java:44]: I think that we'll need a better name for this

Comment (ctron) [reply extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java:44]: Sure, do you have suggestions?

**Comment by ctron** [comment]
> Can you add a description of what exactly the type expansion is doing and why it's needed? Perhaps with an example?

I absolutely understand. Do you want this description in the GitHub comments? Or in the code itself?

File: extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java (hunk 1 above)
Comment (geoand) [reply extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java:44]: Maybe `EXCLUDED_WATCH_CLASSES`?

**Comment by geoand** [comment]
I'd say lets start with a GH comment and see :)

**Comment by ctron** [comment]
The problem is: the kubernetes extension finds the main class, used in the watcher (or informer) and enables this class for the reflection mechanism. So when the kubernetes client pulls in some JSON from the server, it can deserialize the JSON, with that class information. Which is done by Jackson, using reflection.

All attributes that are on this specific class are covered. However, if you inherit fields from another class, or reference some other class, then these attribute cannot be found via reflection. Annotating them with `@RegisterForReflection` did fix issue.

The "expand" functionality (and I am open for a different name) takes the root class, extracted from the Watcher implementation, and walks through the type hierarchy, and also walks through the types of the the fields and getters, adding all referenced types to the reflection system.

Assuming you have the following structure:

~~~java
class AbstractHasMetadata implements HasMetadata {
  String kind;
  String apiVersion;
}

class MyCrd extends AbstractHasMetadata {
  MyCrdSpec spec;
}

class MyCrdSpec {
  List<MyCrdConfig> allTheConfigs;
}

class MyCrdConfig {
  String value;
}

void startWatcher() {
  client.crd(MyCrd.class).watch(new Watcher<>() {
    public void onAdd() {}
    public void onRemove() {}
    public void onUpdate() {}
});
}
~~~

The extension currently detects `MyCrd` and registers that. But doesn't care about the other classes in the type hierarchy/context.

So deserialization fails, when it e.g. finds the field "kind" from the abstract base class. Or when it finds the `allTheConfigs` field.

Now the extensions walks the type hierarchy up, adding the abstract class as well. But it also walks the fields of each type it processes, and finds the type `MyCrdSpec` and `MyCrdConfig`.

**Comment by geoand** [comment]
Now I get it, thanks for the explanation @ctron. 

Last question: Instead of doing the manual walking of the hierarchy, did you try producing `ReflectiveHierarchyBuildItem` for the classes identified for watcher and informer?
I believe it pretty much covers what you did manually

**Comment by ctron** [comment]
> Last question: Instead of doing the manual walking of the hierarchy, did you try producing `ReflectiveHierarchyBuildItem` for the classes identified for watcher and informer?
> I believe it pretty much covers what you did manually

No I didn't, because I wasn't ware of that :grinning: …  I will take a look and give it a try.

**Comment by geoand** [comment]
> No I didn't, because I wasn't ware of that  … I will take a look and give it a try.

Yeah sorry, that's my fault that I didn't bring it to your attention :(

Let me know how it goes

**Comment by ctron** [comment]
Works as expected. I updated the PR.

**Review by geoand — APPROVED** [review]
LGTM, thanks!

**Pull request merged**